ratik/src/intTest/java/org/opencadc/ratik/InventoryValidatorTest.java

```diff
@@ -0,0 +1,759 @@
+        Assert.assertTrue("artifact does not contains remote site location",
+                          localArtifact.siteLocations.contains(new SiteLocation(remoteSiteID)));
+    }
+
+    /**
+     * discrepancy: artifact.uri in both && artifact.id mismatch
+     *
+     * explanation1: same ID collision due to race condition that metadata-sync has to handle
+     * evidence: no more evidence needed
+     * action: pick winner, create DeletedArtifactEvent for loser, delete loser if it is in L
+     *
+     * winner always in remote?
+     * before: Artifact.uri in L && R, L Artifact.id != R Artifact.id
+     * after:  Artifact not in L, DeletedArtifactEvent in L
+     */
+    @Test
+    public void artifactUriCollision_LocalIsStorage() throws Exception {
+        artifactUriCollision(false);
+    }
+
+    @Test
+    public void artifactUriCollision_LocalIsGlobal() throws Exception {
+        artifactUriCollision(true);
+    }
+
+    public void artifactUriCollision(boolean trackSiteLocations) throws Exception {
+        URI contentCheckSum = TestUtil.getRandomMD5();
+        UUID localID = UUID.randomUUID();
+        Artifact localArtifact = new Artifact(localID, URI.create("cadc:INTTEST/one.ext"), contentCheckSum,
+                                              new Date(), 1024L);
+        this.localEnvironment.artifactDAO.put(localArtifact);
+
+        UUID remoteID = UUID.randomUUID();
+        Artifact remoteArtifact = new Artifact(remoteID, URI.create("cadc:INTTEST/one.ext"), contentCheckSum,
+                                               new Date(), 1024L);
+        this.remoteEnvironment.artifactDAO.put(remoteArtifact);
+
+        try {
+            System.setProperty("user.home", TMP_DIR);
+            InventoryValidator testSubject = new InventoryValidator(this.localEnvironment.daoConfig, TestUtil.LUSKAN_URI,
+                                                                    new IncludeArtifacts(),null,
+                                                                    trackSiteLocations);
+            testSubject.run();
+        } finally {
+            System.setProperty("user.home", USER_HOME);
+        }
+
+        localArtifact = this.localEnvironment.artifactDAO.get(localArtifact.getID());
+        Assert.assertNull("local artifact found", localArtifact);
+
+        DeletedArtifactEvent deletedArtifactEvent = this.localEnvironment.deletedArtifactEventDAO.get(localID);
+        Assert.assertNotNull("DeletedArtifactEvent not found", deletedArtifactEvent);
+    }
+
+    /**
+     * discrepancy: artifact in both && valid metaChecksum mismatch
+     *
+     * explanation1: pending/missed artifact update in L
+     * evidence: ??
+     * action: put Artifact
+     *
+     * explanation2: pending/missed artifact update in R
+     * evidence: ??
+     * action: do nothing
+     *
+     * before: Artifact in L & R, L metaChecksum != R metaChecksum
+     * after: Artifact in L with R metaChecksum
+     */
+
```

Comment:
I assume you were going to add two `@Test` methods that call this - it seems like a good idea to run these site-agnostic tests once for each type anyway (as above).

For this discrepancy, the `evidence: ??` for both explanations means we can't tell which is right. Unfortunately, here the actions are known and different so we still need to figure out what evidence can be used to resolve this. Right now I can't recall if I just forgot to go back and fill that in or if I thought about it and couldn't come up with a plan. Since I'll have to think about it again that's equivalent to having forgot to do it :smile: 

So, on the test setup:
* the fact that you create two artifacts (literally `new Artifact`) means that you have setup a collision and that resolution takes precedence over this one, which is about updates to existing artifacts -- in other tests where you created one Artifact and put it both environments it was actually important because they have the same Entity.id and that is what makes them "the same Artifact" (*not* having the same Artifact.uri)
* manually setting the metaChecksum in the localEnvironment makes that record in the database invalid, so that's a bad idea that you don't need

I think separate methods for the two explanations are needed here. For explanation1, I would create an Artifact and put it into localEnvironment and remoteEnvironment. Then modify the Artifact and put it to remote again. Now you have an update in R that has not been applied to L. Modify setup for explanation2.

This discrepancy could have been phrased as "pick a winner" with "no more evidence needed" and the  the action is "pick a winner, ...". For now, get the test setup right and I'll modify the document so that it is expressed that way or I'll modify the "collision" discrepancy so it has two explanations.

Reply:
Yes forgot to add storage and global test. Separated this test into a test for each explanation.